In [95]:
import pandas as pd
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords


In [96]:
df=pd.read_csv('IA2.csv',header=None)
df.set_index(df.columns[0],inplace=True)
df.head()

1
0                                                   
1  This is the place to go after this branch to t...
2  The direct delivery from Star Theater in Chian...
3  It is very delicious food. Good taste Unspoilt...
4  This house is the second time. I used to be a ...
5  Featured is majestic. Located on the road 107 ...

In [97]:
# 1) Tokenize each review in the collection; 

all_tokens=[]
for i in df[1]:
    tokens = word_tokenize(i.lower())
    all_tokens.append(tokens)
print(all_tokens)


[['this', 'is', 'the', 'place', 'to', 'go', 'after', 'this', 'branch', 'to', 'the', 'afternoon', ',', 'not', 'much', 'people', '.', 'now', 'the', 'popular', 'menu', 'is', 'ice', 'crystal', '.', 'hes', 'not', 'long', 'feel', 'very', 'sorry', 't', '^', 't', 'this', 'is', 'a', 'full', 'set', 'of', 'chocolate', 'mud', 'brownie', '(', '155.-', '฿', ')', 'chocolate', 'brownie', 'sprinkle', 'with', 'vanilla', 'ice', 'cream', 'and', 'whipped', 'cream', '.', 'this', 'is', 'a', 'very', 'hot', 'menu', '.', '5/5', 'go', 'to', 'ferrero', 'honey', 'toast', '(', '245.-', '฿', ')', 'honey', 'toro', 'topped', 'with', 'nuts', 'topping', 'up', 'is', 'ferrero', '.', 'fan', 'club', 'we', 'are', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', '😌', 'this', 'is', 'a', 'good', 'service', '.', 'it', '&', '#', '39', ';', 's', 'a', 'little', 'car', 'park', '.', 'if', 'the', 'parking', 'lot', 'of', 'the', 'building', 'will', 'be', 'stamped', 'the', 'first', '2', 'hrs', '20', 'b

In [98]:
#2) Use the tokenized reviews after step 1, lemmatize all the words;

lemmatizer = nltk.stem.WordNetLemmatizer()
lemmatized_tokens = []
for tokens in all_tokens:
    lemmatized_tokens.append([lemmatizer.lemmatize(token) for token in tokens if token.isalpha()])
print(lemmatized_tokens)


[['this', 'is', 'the', 'place', 'to', 'go', 'after', 'this', 'branch', 'to', 'the', 'afternoon', 'not', 'much', 'people', 'now', 'the', 'popular', 'menu', 'is', 'ice', 'crystal', 'he', 'not', 'long', 'feel', 'very', 'sorry', 't', 't', 'this', 'is', 'a', 'full', 'set', 'of', 'chocolate', 'mud', 'brownie', 'chocolate', 'brownie', 'sprinkle', 'with', 'vanilla', 'ice', 'cream', 'and', 'whipped', 'cream', 'this', 'is', 'a', 'very', 'hot', 'menu', 'go', 'to', 'ferrero', 'honey', 'toast', 'honey', 'toro', 'topped', 'with', 'nut', 'topping', 'up', 'is', 'ferrero', 'fan', 'club', 'we', 'are', 'this', 'is', 'a', 'good', 'service', 'it', 's', 'a', 'little', 'car', 'park', 'if', 'the', 'parking', 'lot', 'of', 'the', 'building', 'will', 'be', 'stamped', 'the', 'first', 'hr', 'baht', 'sit', 'a', 'long', 'time', 'but', 'if', 'you', 'park', 'other', 'building', 'it', 'will', 'be', 'but', 'it', 's', 'just', 'this', 'parking', 'lot', 'is', 'already', 'low'], ['the', 'direct', 'delivery', 'from', 'star',

In [99]:
#3)  Based on the output in step 2, remove all the stop‐words and the punctuations;
#----Removed punctuation in last step when checking for isalpha.

stop_words = set(stopwords.words('english'))
stop_words_removed = [[t for t in review if not t in stop_words] for review in lemmatized_tokens]

#print(stop_words_removed)

In [100]:
docs = [' '.join(tokens) for tokens in stop_words_removed]
print(docs)

['place go branch afternoon much people popular menu ice crystal long feel sorry full set chocolate mud brownie chocolate brownie sprinkle vanilla ice cream whipped cream hot menu go ferrero honey toast honey toro topped nut topping ferrero fan club good service little car park parking lot building stamped first hr baht sit long time park building parking lot already low', 'direct delivery star theater chiang mai guarantee really cool variety flavor choose baht per piece taste choose buy buy', 'delicious food good taste unspoilt people like eat spicy much cooked shop ha prepared chili sauce chili sauce tomato service ok', 'house second time used pig pan seafood price baht mall use kind common sea food fresh diarrhea first time eat diarrhea cockroach stink rotten eat shrimp enough review', 'featured majestic located road mae taeng restaurant cozy atmosphere today people much green tea quietly enough dessert good enough bring piece high price place traveler stay mont cafe buy leave frien

In [101]:
# 4)  Based on the output in step 3, convert each of the reviews in TD‐IDF vectors. The minimal 
# document frequency for each term is 3. Also, include 2‐gram; 

###TF-IDF
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer1 = TfidfVectorizer(ngram_range=(1,2), min_df=3)
vectorizer1.fit(docs)
print(vectorizer1.vocabulary_)

v1 = vectorizer1.transform(docs)
td_idf=v1.toarray()
df_new = pd.DataFrame(td_idf,columns=vectorizer1.get_feature_names())
df_new = df_new.reset_index(drop=True)
df_new.index = df_new.index + 1

print(df_new)
print(df_new.shape)

{'place': 842, 'go': 460, 'branch': 118, 'much': 738, 'people': 819, 'popular': 852, 'menu': 712, 'ice': 553, 'long': 653, 'feel': 381, 'full': 440, 'set': 1021, 'cream': 230, 'whipped': 1279, 'hot': 539, 'honey': 531, 'fan': 373, 'good': 470, 'service': 1020, 'little': 645, 'car': 140, 'park': 804, 'parking': 805, 'lot': 662, 'building': 131, 'first': 404, 'baht': 71, 'sit': 1049, 'time': 1185, 'already': 26, 'low': 669, 'car park': 142, 'long time': 654, 'star': 1098, 'theater': 1169, 'chiang': 164, 'mai': 675, 'really': 926, 'cool': 219, 'variety': 1235, 'flavor': 410, 'choose': 174, 'per': 823, 'piece': 839, 'taste': 1149, 'buy': 133, 'chiang mai': 165, 'delicious': 259, 'food': 416, 'like': 634, 'eat': 302, 'spicy': 1089, 'shop': 1029, 'ha': 492, 'chili': 169, 'sauce': 984, 'ok': 772, 'good taste': 471, 'people like': 821, 'like eat': 635, 'house': 541, 'second': 1000, 'used': 1231, 'pig': 840, 'pan': 799, 'seafood': 997, 'price': 864, 'mall': 686, 'use': 1230, 'kind': 593, 'commo

In [102]:
#concating the new dataframe that has td-idf vectors with original data frame that has the 100 reviews

result = pd.concat([df, df_new], axis=1)
print(result)
print(result.shape)

                                                     1      able  absolutely  \
1    This is the place to go after this branch to t...  0.000000         0.0   
2    The direct delivery from Star Theater in Chian...  0.000000         0.0   
3    It is very delicious food. Good taste Unspoilt...  0.000000         0.0   
4    This house is the second time. I used to be a ...  0.000000         0.0   
5    Featured is majestic. Located on the road 107 ...  0.000000         0.0   
..                                                 ...       ...         ...   
96   Perhaps, the old people in this film do not re...  0.000000         0.0   
97   There are many handsome guys, many beautiful w...  0.000000         0.0   
98   The first half is still okay, it seems to be m...  0.000000         0.0   
99   Like most of the mango-produced thunder dramas...  0.100338         0.0   
100  Some of Liu Zhenyun’s works have been read int...  0.000000         0.0   

       accept  accident  accompanied  a

In [103]:
#Saving to csv

result.to_csv('Part1_Q4.csv', index=False)


In [104]:
# 5)  Based on the output in step 1, POS‐tag each word and do a TD‐IDF vectorization, the minimal 
#document frequency for each term is 4 (please don’t do normalization and stop‐word removal)

POS_c2 = []
for doc in all_tokens:
    word_list=[]
    for token in doc:
        if token.isalpha():
            word_list.append(token)
    #token_doc = nltk.word_tokenize(doc)
    POS_token_doc = nltk.pos_tag(word_list)
    POS_token_temp = []
    for i in POS_token_doc:
         POS_token_temp.append(i[0] + "_" + i[1])
    POS_c2.append(" ".join(POS_token_temp))         
    
print(POS_c2)

['this_DT is_VBZ the_DT place_NN to_TO go_VB after_IN this_DT branch_NN to_TO the_DT afternoon_NN not_RB much_JJ people_NNS now_RB the_DT popular_JJ menu_NN is_VBZ ice_JJ crystal_JJ hes_NNS not_RB long_RB feel_VB very_RB sorry_JJ t_NN t_NN this_DT is_VBZ a_DT full_JJ set_NN of_IN chocolate_NN mud_NN brownie_NN chocolate_NN brownie_NN sprinkle_NN with_IN vanilla_NN ice_NN cream_NN and_CC whipped_VBD cream_NN this_DT is_VBZ a_DT very_RB hot_JJ menu_NN go_VBP to_TO ferrero_VB honey_NN toast_NN honey_NN toro_NN topped_VBD with_IN nuts_NNS topping_VBG up_RP is_VBZ ferrero_JJ fan_NN club_NN we_PRP are_VBP this_DT is_VBZ a_DT good_JJ service_NN it_PRP s_VBZ a_DT little_JJ car_NN park_NN if_IN the_DT parking_NN lot_NN of_IN the_DT building_NN will_MD be_VB stamped_VBN the_DT first_JJ hrs_NN baht_NN sit_VBD a_DT long_JJ time_NN but_CC if_IN you_PRP park_VBP other_JJ buildings_NNS it_PRP will_MD be_VB but_CC it_PRP s_VBZ just_RB this_DT parking_NN lot_NN is_VBZ already_RB low_JJ', 'the_DT direct

In [105]:
###TF-IDF
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer2 = TfidfVectorizer(min_df=4)
vectorizer2.fit(POS_c2)
print(vectorizer2.vocabulary_)
v2 = vectorizer2.transform(POS_c2)
pos=v2.toarray()

pos_df = pd.DataFrame(pos,columns=vectorizer2.get_feature_names())
pos_df = pos_df.reset_index(drop=True)
pos_df.index = pos_df.index + 1

{'this_dt': 793, 'is_vbz': 393, 'the_dt': 777, 'place_nn': 579, 'to_to': 801, 'go_vb': 314, 'after_in': 13, 'not_rb': 518, 'much_jj': 498, 'people_nns': 566, 'now_rb': 522, 'popular_jj': 592, 'menu_nn': 482, 'long_rb': 437, 'feel_vb': 257, 'very_rb': 840, 't_nn': 758, 'a_dt': 0, 'full_jj': 298, 'set_nn': 687, 'of_in': 526, 'with_in': 878, 'ice_nn': 375, 'cream_nn': 158, 'and_cc': 27, 'hot_jj': 362, 'go_vbp': 315, 'up_rp': 826, 'fan_nn': 252, 'we_prp': 860, 'are_vbp': 33, 'good_jj': 320, 'service_nn': 686, 'it_prp': 394, 's_vbz': 653, 'little_jj': 433, 'car_nn': 102, 'if_in': 378, 'lot_nn': 445, 'building_nn': 93, 'will_md': 876, 'be_vb': 55, 'first_jj': 274, 'baht_nn': 52, 'long_jj': 436, 'time_nn': 798, 'but_cc': 94, 'you_prp': 901, 'other_jj': 546, 'just_rb': 401, 'already_rb': 20, 'low_jj': 453, 'from_in': 294, 'star_nn': 731, 'theater_nn': 778, 'in_in': 383, 'i_nn': 369, 'that_in': 775, 'really_rb': 626, 'cool_jj': 152, 'there_ex': 783, 'variety_nn': 835, 'choose_vb': 121, 'per_in'

In [106]:
#concating the new dataframe that has td-idf vectors for pos tag with original data frame that has the 100 reviews

result_pos = pd.concat([df, pos_df], axis=1)
print(result_pos)
print(result_pos.shape)

                                                     1      a_dt   able_jj  \
1    This is the place to go after this branch to t...  0.208217  0.000000   
2    The direct delivery from Star Theater in Chian...  0.067085  0.000000   
3    It is very delicious food. Good taste Unspoilt...  0.000000  0.000000   
4    This house is the second time. I used to be a ...  0.117817  0.000000   
5    Featured is majestic. Located on the road 107 ...  0.110532  0.000000   
..                                                 ...       ...       ...   
96   Perhaps, the old people in this film do not re...  0.165949  0.000000   
97   There are many handsome guys, many beautiful w...  0.090385  0.000000   
98   The first half is still okay, it seems to be m...  0.117429  0.000000   
99   Like most of the mango-produced thunder dramas...  0.190971  0.052935   
100  Some of Liu Zhenyun’s works have been read int...  0.151063  0.000000   

     about_in  above_in  accident_nn  acting_vbg  actor_nn  act

In [107]:
#Saving to csv


result_pos.to_csv('Part1_Q5.csv', index=False)


In [108]:
# Use a markdown at the end of the .ipynb file 
# to report the dimension of the vectors of step 4 and step 5. 

In [109]:
print(td_idf.shape) # dimension of tf-idf vector from step 4

(100, 1322)


In [110]:
print(pos.shape) # dimension of tf-idf vector from step 5

(100, 906)


Dimension of vector from step 4 is (100, 1322) and the dimension from step 5 is (100, 906)